File: src/main/java/com/woomulwoomul/woomulwoomulbackend/config/auth/CustomAuthenticationEntryPoint.java
package com.woomulwoomul.woomulwoomulbackend.config.auth;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.woomulwoomul.woomulwoomulbackend.common.response.ExceptionCode;
import com.woomulwoomul.woomulwoomulbackend.common.response.ExceptionResponse;
import com.woomulwoomul.woomulwoomulbackend.exception.CustomException;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.OutputStream;

import static com.woomulwoomul.woomulwoomulbackend.common.response.ExceptionCode.SERVER_ERROR;
import static com.woomulwoomul.woomulwoomulbackend.common.response.ExceptionCode.TOKEN_UNAUTHENTICATED;
import static org.springframework.http.HttpStatus.UNAUTHORIZED;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@Component
@RequiredArgsConstructor
public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {

    private final ObjectMapper objectMapper;

    @Override
    public void commence(HttpServletRequest request,
                         HttpServletResponse response,
                         AuthenticationException authException) throws IOException, ServletException {
        response.setStatus(UNAUTHORIZED.value());
        response.setContentType(APPLICATION_JSON_VALUE);

        try {
            String responseBody = objectMapper.writeValueAsString(ExceptionResponse.ofResponseEntity(TOKEN_UNAUTHENTICATED));

            OutputStream outputStream = response.getOutputStream();
            outputStream.write(responseBody.getBytes());
            outputStream.flush();
        } catch (IOException e) {
            throw new CustomException(SERVER_ERROR, e.getCause());
        }
    }
}

File: src/main/kotlin/com/woomulwoomul/woomulwoomulbackend/config/auth/CustomAuthenticationEntryPoint.kt
package com.woomulwoomul.woomulwoomulbackend.config.auth

import com.fasterxml.jackson.databind.ObjectMapper
import com.woomulwoomul.woomulwoomulbackend.common.constant.ExceptionCode.*
import com.woomulwoomul.woomulwoomulbackend.common.response.CustomException
import com.woomulwoomul.woomulwoomulbackend.common.response.ExceptionResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.stereotype.Component
import java.io.IOException

@Component
class CustomAuthenticationEntryPoint(
    private val objectMapper: ObjectMapper,
) : AuthenticationEntryPoint {

    override fun commence(request: HttpServletRequest?,
                          response: HttpServletResponse?,
                          authException: AuthenticationException?) {
        response?.apply {
            status = HttpStatus.UNAUTHORIZED.value()
            contentType = APPLICATION_JSON_VALUE
            try {
                val responseBody = objectMapper.writeValueAsString(ExceptionResponse(TOKEN_UNAUTHENTICATED))
                outputStream.write(responseBody.toByteArray())
            } catch (e: IOException) {
                throw CustomException(SERVER_ERROR, e.cause)
            }
            outputStream.flush()
        }
    }
}